library/std/src/lazy.rs

```diff
@@ -348,22 +342,12 @@ impl<T> SyncOnceCell<T> {
     /// ```
     #[unstable(feature = "once_cell", issue = "74465")]
     pub fn take(&mut self) -> Option<T> {
-        mem::take(self).into_inner()
-    }
-
-    /// Takes the wrapped value out of a `SyncOnceCell`.
-    /// Afterwards the cell is no longer initialized.
-    ///
-    /// Safety: The cell must now be free'd WITHOUT dropping. No other usages of the cell
-    /// are valid. Only used by `into_inner` and `drop`.
-    unsafe fn take_inner(&mut self) -> Option<T> {
-        // The mutable reference guarantees there are no other threads that can observe us
-        // taking out the wrapped value.
-        // Right after this function `self` is supposed to be freed, so it makes little sense
-        // to atomically set the state to uninitialized.
         if self.is_initialized() {
-            let value = mem::replace(&mut self.value, UnsafeCell::new(MaybeUninit::uninit()));
-            Some(value.into_inner().assume_init())
+            self.once = Once::new();
+            // SAFETY: `self.value` is initialized and contains a valid `T`.
+            // `self.once` is reset, so `is_initialized()` will be false again
+            // which prevents the value from being read twice.
+            unsafe { Some((&mut *self.value.get()).assume_init_read()) }
```

Comment:
So we leave some garbage bytes inside, but they are guarded by reseted `Once`? Nifty!

Reply:
FWIW, previous this wrote `MaybeUninit::uninit()` which is equivalent to not writing anything (just the compiler might find it easier this way as there are no uninit writes to eliminate).

Reply:
The disassembly of this new version is slightly shorter. It seems to make one less copy. (Specifically for `SyncOnceCell<String>` at least.) But that's probably just the copy *out* of `self.value` into `value`.

Reply:
It's also possible that the compiler just did not realize before that it was needlessly copying uninitialized memory.